### DISCUSSION TITLE
PlatformIO problem working with SD card

### DISCUSSION THREAD

**Discussion by MHYCWasTaken** [Bodmer/TFT_eSPI#2737] [General]
PlatformIO, esp32
---
I am using a touch screen(ILI9341), the screen and touch worked well  
But `SD.begin()` gives me  
```
[   402][E][esp32-hal-cpu.c:110] addApbChangeCallback(): duplicate func=0x400d8098 arg=0x3ffbdc60
[  1916][E][sd_diskio.cpp:126] sdSelectCard(): Select Failed
[  2416][E][sd_diskio.cpp:126] sdSelectCard(): Select Failed
[  2416][E][sd_diskio.cpp:802] sdcard_mount(): f_mount failed: (3) The physical drive cannot work
[  2948][E][sd_diskio.cpp:126] sdSelectCard(): Select Failed
```
And `SD.open()` gives me  
```
[  2948][E][vfs_api.cpp:24] open(): File system is not mounted
```
My code is like:
``` cpp
#include <arduino.h>
#include <SPI.h>
#include <TFT_eSPI.h>
#include <SD.h>
TFT_eSPI tft = TFT_eSPI();
...
void init() {
  ...
  tft.init();
  pinMode(12, OUTPUT);
  SD.begin(12, SPI, 1000000);
  File f = SD.open("test.txt", FILE_WRITE);
  ...
}
...
```
I didn't do `SPI.begin()`, so im kinda sure SD and TFT are using a same SPI object (that `tft.init()` created)  
But when I add `SPI.begin(18, 19, 23);` between `tft.init()` and `SD.init()`
(There should be 4 args, the last arg is `ss`, but i dont know what `ss`is , is it same as `cs`? if it is, should i use SD's cs or TFT's cs or TOUCH's cs?)
`SPI.begin()` gives me
```
[   402][E][esp32-hal-cpu.c:110] addApbChangeCallback(): duplicate func=0x400d80a8 arg=0x3ffbdc60
```
`SD.begin()` gives me
```
[   720][E][sd_diskio.cpp:199] sdCommand(): Card Failed! cmd: 0x37
[   720][E][sd_diskio.cpp:199] sdCommand(): Card Failed! cmd: 0x29
[   727][E][sd_diskio.cpp:802] sdcard_mount(): f_mount failed: (3) The physical drive cannot work
```
And `SD.open()` give the same error
```
[   959][E][vfs_api.cpp:24] open(): File system is not mounted
```
Then, I put `SPI.begin()` line in front of `tft.begin()`, The error is as same as that when i didn't add `SPI.begin()`  
(So i guess `tft.begin()` overwrites SPI object?)

My wire connection is like
```
TOUCH_MISO & SD_MISO ---> 19
TFT_MOSI & TOUCH_MOSI & SD_MOSI ---> 23 
TFT_SCLK & TOUCH_SLCK & SD_SLCK ---> 18
TFT_CS ---> 15
TOUCH_CS ---> 21
SD_CS ---> 12
```

My SD card is 32gb, but i have just make a partition with 1.5gb size and FAT16, the left space has no partition  
Accordint to [1509](https://github.com/Bodmer/TFT_eSPI/issues/1509), I have no idea im using which of FSPI, VSPI or HSPI, i have no code setting them  

---
(sorry my english is a little poor)
(and im basiclly new to hardware stuff)
---
By the way, ive post this question to [1509](https://github.com/Bodmer/TFT_eSPI/issues/1509), but there was no reply  
Should i not post question to a closed issue?

**Comment by Zombieschannel** [comment]
Maybe try to change the SD.begin(12, SPI, 1000000); to SD.begin(12, tft.getSPIinstance(), 1000000); ?
Also, ss is the same as cs afaik.

**Reply by MHYCWasTaken** [reply]
Hi  
Not working, I got the same error  
as
```
[  1604][E][sd_diskio.cpp:126] sdSelectCard(): Select Failed
[  2104][E][sd_diskio.cpp:126] sdSelectCard(): Select Failed
[  2104][E][sd_diskio.cpp:199] sdCommand(): Card Failed! cmd: 0x29
[  2104][E][sd_diskio.cpp:802] sdcard_mount(): f_mount failed: (3) The physical drive cannot work
[  2707][E][sd_diskio.cpp:126] sdSelectCard(): Select Failed
```
Is there any way to make tft use a specific spi object?

**Reply by Zombieschannel** [reply]
Not sure if there is a way to make tft use a specific object. Not sure if you know this but esp32 has 2 SPIs. VSPI and HSPI. You could use one for SD card and one for tft(User config in the tft library has an option to use HSPI instead of VSPI). I know I got the sd + tft to work on either both SPIs or on a single VSPI, so not sure why this wouldn't work.

**Comment by TangerineDreamer** [comment]
try in the setup:  
```
SD.begin(PIN_SD_CARD_CLK, PIN_SD_CARD_MISO, PIN_SD_CARD_MOSI, PIN_SD_CARD_CS);
```
then:
```
 if (!sd.begin(12,10)) {
    Serial.println("SD card error!");
    while (true);
  }
```
etc...

**Reply by MHYCWasTaken** [reply]
Sorry, I didn't really get your point  
Where comes `sd` ? Shouldn't it be `SD`?  
And the second arg of `SD.begin()` should be a `SPIClass`  
But I tried this code anyway:
``` cpp
#include <arduino.h>
#include <SPI.h>
#include <TFT_eSPI.h>
#include <SD.h>
TFT_eSPI tft = TFT_eSPI();
...
void init() {
  ...
  tft.init();
  pinMode(12, OUTPUT);
  Serial.println("1");
  SPI.begin(18, 19, 23, 12);
  Serial.println("2");
  if(! SD.begin(12)) { // SD.begin(12, SPI) gives me same error
    Serianl.println("E");
    while(1);
  }
  File f = SD.open("test.txt", FILE_WRITE);
  ...
}
...
```
it gives me:
```
1
[   402][E][esp32-hal-cpu.c:110] addApbChangeCallback(): duplicate func=0x400d7fb8 arg=0x3ffbdbfc
2
[   705][E][sd_diskio.cpp:199] sdCommand(): Card Failed! cmd: 0x37
[   705][E][sd_diskio.cpp:199] sdCommand(): Card Failed! cmd: 0x29
[   712][E][sd_diskio.cpp:802] sdcard_mount(): f_mount failed: (3) The physical drive cannot work
E
```
If i use `SD.begin(12, tft.getSPIinstance())`, it gives me two more lines:
```
  1
  [   402][E][esp32-hal-cpu.c:110] addApbChangeCallback(): duplicate func=0x400d7fc4 arg=0x3ffbdbfc
  2
+ [  1211][E][sd_diskio.cpp:126] sdSelectCard(): Select Failed
  [  1211][E][sd_diskio.cpp:199] sdCommand(): Card Failed! cmd: 0x37
  [  1211][E][sd_diskio.cpp:199] sdCommand(): Card Failed! cmd: 0x29
  [  1282][E][sd_diskio.cpp:802] sdcard_mount(): f_mount failed: (3) The physical drive cannot work
+ [  1885][E][sd_diskio.cpp:126] sdSelectCard(): Select Failed
  E
```
Since i did `SPI.begin()` with SD's CS, i've also tried `SD.begin()` with no args:
```
1
[   402][E][esp32-hal-cpu.c:110] addApbChangeCallback(): duplicate func=0x400d7fb8 arg=0x3ffbdbfc
2
[  1715][E][sd_diskio.cpp:126] sdSelectCard(): Select Failed
[  2215][E][sd_diskio.cpp:126] sdSelectCard(): Select Failed
[  2215][E][sd_diskio.cpp:199] sdCommand(): Card Failed! cmd: 0x29
[  2215][E][sd_diskio.cpp:802] sdcard_mount(): f_mount failed: (3) The physical drive cannot work
[  2818][E][sd_diskio.cpp:126] sdSelectCard(): Select Failed
E
```

**Comment by TangerineDreamer** [comment]
ooops sorry,i was dealing with sdfat,but with SD.h the name is SD

```
#include <arduino.h>
#include <SPI.h>
#include <TFT_eSPI.h>
#include <SD.h>
TFT_eSPI tft = TFT_eSPI();

// **************** SD CARD
const int SD_CS_PIN = 12; //your cs pin
#define SPI_SPEED SD_SCK_MHZ(10) //around this

etc........

void setup() {
Serial.begin(115200);
SPI.begin(18, 19, 23, 12);

if (!SD.begin(SD_CS_PIN, SPI_SPEED)) { //the second argument is the speed in Mhz ,10 to 15 is the deal
    Serial.println("SD card error!");
    while (true);
  }
etc...
```

**Reply by MHYCWasTaken** [reply]
![图片](https://github.com/Bodmer/TFT_eSPI/assets/89076764/b5307728-1318-40d3-8bf0-86cdfe0af7f3)
As you see, the second parameter should be a SPIClass and it's default `SPI` , the third parameter is the SPI_SPEED  
So I tried `SD.begin(12, SPI, 10)`, And it didn't work  
```
1
[   402][E][esp32-hal-cpu.c:110] addApbChangeCallback(): duplicate func=0x400d7fb8 arg=0x3ffbdbfc
2
[  1320][E][sd_diskio.cpp:199] sdCommand(): Card Failed! cmd: 0x37
[  1320][E][sd_diskio.cpp:199] sdCommand(): Card Failed! cmd: 0x29
[  1327][E][sd_diskio.cpp:802] sdcard_mount(): f_mount failed: (3) The physical drive cannot work
[  1930][E][sd_diskio.cpp:126] sdSelectCard(): Select Failed
E
```
Well im thinking about if this is a problem with my sd card  
The card is a little old, but my computer can read it perfectly  
So I've bought another sd card online

**Reply by ffkarine** [reply]
Hey @MHYCWasTaken ! 

Were you able to solve this error? 
I'm not using this library, but I'm facing the same issue when trying to connect a W5500 module in a SPI interface and a SD module in another SPI interface.

**Comment by TangerineDreamer** [comment]
That's weird!!

Here : https://www.arduino.cc/reference/en/libraries/sd/begin/
it's written "SD.begin(cspin)" where cspin is optionnal.

Someone wrote here: https://arduino.stackexchange.com/questions/33455/why-is-sd-card-failing-to-begin
       "  I only initialize the card once at "sd.cardBegin(SD_CHIP_SELECT,  SPI_HALF_SPEED)",  "

I can tell you i also had problems with "sdcard CS failed" all of because it was a 32Gb card and the one wich worked
was a 16Gb,maybe it's the source of your problem.

If you use a esp32 devkit1 (regular one),take a look at this post:
https://github.com/Bodmer/TFT_eSPI/discussions/2717

The example i posted there works with a sdcard and a separate tft, but use HSPI & VSPI,don't need to activate
anything with "digitalWrite procedure" in the setup,it works fine as it is.
